File: android/app/src/main/kotlin/com/stillmoment/presentation/viewmodel/TimerViewModel.kt
package com.stillmoment.presentation.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.stillmoment.domain.models.IntervalSettings
import com.stillmoment.domain.models.MeditationSettings
import com.stillmoment.domain.models.Praxis
import com.stillmoment.domain.models.TimerAction
import com.stillmoment.domain.models.TimerEffect
import com.stillmoment.domain.models.TimerEvent
import com.stillmoment.domain.models.TimerState
import com.stillmoment.domain.repositories.PraxisRepository
import com.stillmoment.domain.repositories.SoundCatalogRepository
import com.stillmoment.domain.repositories.TimerRepository
import com.stillmoment.domain.services.AudioServiceProtocol
import com.stillmoment.domain.services.SoundscapeResolverProtocol
import com.stillmoment.domain.services.TimerForegroundServiceProtocol
import com.stillmoment.domain.services.TimerReducer
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

/**
 * ViewModel managing timer state and user interactions.
 *
 * Uses Unidirectional Data Flow (UDF) with TimerReducer for pure state transitions
 * and effect handling for side effects (audio, persistence, foreground service).
 *
 * Timer events (preparation completed, meditation completed, interval gong due) are
 * emitted by `MeditationTimer.tick()` as domain events instead of being detected
 * via previousState comparison.
 */
@Suppress("TooManyFunctions") // ViewModel naturally has many user-facing action methods
@HiltViewModel
class TimerViewModel
@Inject
constructor(
    application: Application,
    private val timerRepository: TimerRepository,
    private val audioService: AudioServiceProtocol,
    private val foregroundService: TimerForegroundServiceProtocol,
    private val praxisRepository: PraxisRepository,
    private val soundCatalogRepository: SoundCatalogRepository,
    private val soundscapeResolver: SoundscapeResolverProtocol
) : AndroidViewModel(application) {
    private val _uiState = MutableStateFlow(TimerUiState())
    val uiState: StateFlow<TimerUiState> = _uiState.asStateFlow()

    private var timerJob: Job? = null

    /** Returns the display name for the background sound (built-in or custom), or null if silent. */
    private fun resolveBackgroundSoundName(praxis: Praxis): String? {
        return soundscapeResolver.resolve(praxis.backgroundSoundId)?.displayName
    }

    init {
        // Load initial settings and praxis synchronously (DataStore is fast)
        // This ensures the UI shows the saved duration immediately, like iOS with UserDefaults
        val initialPraxis = runBlocking { praxisRepository.load() }
        val initialSettings = initialPraxis.toMeditationSettings()
        _uiState.value = TimerUiState(
            timer = null,
            selectedMinutes = initialSettings.durationMinutes,
            settings = initialSettings,
            currentPraxis = initialPraxis,
            builtInSounds = soundCatalogRepository.getAllSounds(),
            resolvedBackgroundSoundName = resolveBackgroundSoundName(initialPraxis)
        )
        // Observe praxis changes and sync settings + pill labels
        observePraxis()

        // Subscribe to audio completion flows
        viewModelScope.launch {
            audioService.gongCompletionFlow.collect {
                onGongCompleted()
            }
        }
    }

    // MARK: - Action Dispatch (Unidirectional Data Flow)

    /**
     * Dispatches an action through the reducer and handles resulting effects.
     */
    private fun dispatch(action: TimerAction) {
        val current = _uiState.value
        val effects = TimerReducer.reduce(
            action = action,
            timerState = current.timerState,
            selectedMinutes = current.selectedMinutes,
            settings = current.settings,
        )
        effects.forEach { handleEffect(it) }
    }

    /**
     * Handles side effects produced by the reducer.
     */
    @Suppress("CyclomaticComplexMethod") // Effect handler covers all sealed class variants
    private fun handleEffect(effect: TimerEffect) {
        when (effect) {
            is TimerEffect.StartForegroundService -> {
                foregroundService.startService(
                    effect.soundId,
                    effect.soundVolume,
                    effect.gongSoundId,
                    effect.gongVolume
                )
            }
            is TimerEffect.StopForegroundService -> foregroundService.stopService()
            is TimerEffect.PlayStartGong -> foregroundService.playGong(effect.gongSoundId, effect.gongVolume)
            is TimerEffect.PlayIntervalGong -> {
                foregroundService.playIntervalGong(effect.gongSoundId, effect.gongVolume)
            }
            is TimerEffect.PlayCompletionSound -> foregroundService.playGong(effect.gongSoundId, effect.gongVolume)
            is TimerEffect.StartBackgroundAudio -> {
                foregroundService.updateBackgroundAudio(effect.soundId, effect.soundVolume)
            }
            is TimerEffect.StartTimer -> handleStartTimer(effect)
            is TimerEffect.ResetTimer -> handleResetTimer()
            is TimerEffect.TransitionToRunning -> handleTransitionToRunning()
            is TimerEffect.TransitionToCompleted -> handleTransitionToCompleted()
            is TimerEffect.SaveSettings -> viewModelScope.launch {
                val currentPraxis = _uiState.value.currentPraxis
                praxisRepository.save(
                    currentPraxis.withDurationMinutes(effect.settings.durationMinutes)
                )
            }
        }
    }

    private fun handleStartTimer(effect: TimerEffect.StartTimer) {
        viewModelScope.launch {
            val initialEvents = timerRepository.start(
                effect.durationMinutes,
                effect.preparationTimeSeconds
            )
            // Read the initialized timer directly — no tick, no decrement
            timerRepository.currentTimer?.let { initialTimer ->
                _uiState.update { it.copy(timer = initialTimer) }
            }
            processTimerEvents(initialEvents)
        }
        startTimerLoop()
    }

    private fun handleResetTimer() {
        timerJob?.cancel()
        viewModelScope.launch {
            timerRepository.reset()
            _uiState.update { it.copy(timer = null) }
        }
    }

    private fun handleTransitionToRunning() {
        timerRepository.startRunning()
        _uiState.update { state ->
            state.copy(timer = state.timer?.withState(TimerState.Running))
        }
    }

    private fun handleTransitionToCompleted() {
        timerRepository.completeTimer()
        _uiState.update { state ->
            state.copy(timer = state.timer?.withState(TimerState.Completed))
        }
    }

    // MARK: - Public Methods (User Actions)

    fun setSelectedMinutes(minutes: Int) {
        _uiState.update { state ->
            val updatedSettings = state.settings.withDurationMinutes(minutes)
            state.copy(
                selectedMinutes = updatedSettings.durationMinutes,
                settings = updatedSettings
            )
        }
        saveSettings()
    }

    fun startTimer() {
        _uiState.update { state ->
            state.copy(currentAffirmationIndex = (state.currentAffirmationIndex + 1) % AFFIRMATION_COUNT)
        }
        dispatch(TimerAction.StartPressed)
    }

    fun resetTimer() {
        dispatch(TimerAction.ResetPressed)
    }

    fun showSettings() {
        _uiState.update { it.copy(showSettings = true) }
    }

    fun hideSettings() {
        stopGongPreview()
        stopBackgroundPreview()
        _uiState.update { it.copy(showSettings = false) }
    }

    // MARK: - Audio Preview

    /**
     * Play a gong sound preview. Automatically stops any previous preview.
     * Uses the current gong volume setting for preview playback.
     */
    fun playGongPreview(soundId: String) {
        audioService.playGongPreview(soundId, _uiState.value.settings.gongVolume)
    }

    /**
     * Play interval gong sound preview.
     * Uses the current interval gong sound and volume settings for preview playback.
     */
    fun playIntervalGongPreview() {
        val settings = _uiState.value.settings
        audioService.playIntervalGong(settings.intervalSoundId, settings.intervalGongVolume)
    }

    /**
     * Play a specific interval gong sound preview (e.g., when changing sound selection).
     */
    fun playIntervalGongPreview(soundId: String) {
        audioService.playIntervalGong(soundId, _uiState.value.settings.intervalGongVolume)
    }

    /**
     * Stop the current gong preview.
     */
    fun stopGongPreview() {
        audioService.stopGongPreview()
    }

    /**
     * Play a background sound preview. Automatically stops any previous preview (gong or background).
     * Plays for 3 seconds with fade-out.
     * Uses the current volume setting from settings for preview playback.
     */
    fun playBackgroundPreview(soundId: String) {
        audioService.playBackgroundPreview(soundId, _uiState.value.settings.backgroundSoundVolume)
    }

    /**
     * Stop the current background preview.
     */
    fun stopBackgroundPreview() {
        audioService.stopBackgroundPreview()
    }

    // MARK: - Settings Management

    fun updateSettings(settings: MeditationSettings) {
        _uiState.update { it.copy(settings = settings) }
        // Sync selectedMinutes if idle
        if (_uiState.value.timerState == TimerState.Idle) {
            _uiState.update { it.copy(selectedMinutes = settings.durationMinutes) }
        }
        saveSettings()
    }

    fun clearError() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    // MARK: - Timer Loop

    private fun startTimerLoop() {
        timerJob?.cancel()
        timerJob =
            viewModelScope.launch {
                var shouldContinue = true
                while (shouldContinue) {
                    delay(1000L)
                    shouldContinue = processTimerTick()
                }
            }
    }

    /**
     * Builds interval settings from current meditation settings.
     * Returns null when interval gongs are disabled (tick() skips interval detection).
     */
    private fun buildIntervalSettings(): IntervalSettings? {
        val settings = _uiState.value.settings
        if (!settings.intervalGongsEnabled) return null
        return IntervalSettings(
            intervalMinutes = settings.intervalMinutes,
            mode = settings.intervalMode
        )
    }

    /**
     * Processes a single timer tick. Returns true if loop should continue.
     *
     * Domain events from tick() are processed directly — no previousState comparison needed.
     */
    private fun processTimerTick(): Boolean {
        val (updatedTimer, events) = timerRepository.tick(buildIntervalSettings()) ?: return false

        // Update timer directly — no dispatch needed
        _uiState.update { it.copy(timer = updatedTimer) }

        // Process domain events emitted by tick()
        processTimerEvents(events)

        if (updatedTimer.isCompleted) return false

        val activeStates = setOf(
            TimerState.Preparation,
            TimerState.StartGong,
            TimerState.Running
        )
        return updatedTimer.state in activeStates
    }

    /**
     * Processes domain events from [com.stillmoment.domain.models.MeditationTimer.tick] and
     * dispatches corresponding actions.
     */
    private fun processTimerEvents(events: List<TimerEvent>) {
        for (event in events) {
            when (event) {
                TimerEvent.PreparationCompleted -> dispatch(TimerAction.PreparationFinished)
                TimerEvent.MeditationCompleted -> {
                    timerJob?.cancel()
                    dispatch(TimerAction.TimerCompleted)
                }
                TimerEvent.IntervalGongDue -> dispatch(TimerAction.IntervalGongTriggered)
            }
        }
    }

    /**
     * Handles gong audio completion callback.
     * Dispatches the appropriate action based on current timer state:
     * - StartGong -> StartGongFinished (start gong finished, proceed to next phase)
     * - EndGong -> EndGongFinished (completion gong finished, meditation complete)
     */
    private fun onGongCompleted() {
        when (_uiState.value.timerState) {
            TimerState.EndGong -> dispatch(TimerAction.EndGongFinished)
            else -> dispatch(TimerAction.StartGongFinished)
        }
    }

    // MARK: - Persistence

    private fun observePraxis() {
        viewModelScope.launch {
            praxisRepository.praxisFlow.collect { praxis ->
                val settings = praxis.toMeditationSettings()
                val bgName = resolveBackgroundSoundName(praxis)
                _uiState.update { state ->
                    val newMinutes = if (state.timerState == TimerState.Idle) {
                        settings.durationMinutes
                    } else {
                        state.selectedMinutes
                    }
                    state.copy(
                        settings = settings,
                        selectedMinutes = newMinutes,
                        currentPraxis = praxis,
                        resolvedBackgroundSoundName = bgName
                    )
                }
            }
        }
    }

    /**
     * Applies a just-saved Praxis directly to the UI state, bypassing the DataStore roundtrip.
     *
     * Called after PraxisEditor saves so the timer sees the new settings immediately,
     * without racing against the async DataStore write.
     */
    fun applyPraxisUpdate(praxis: Praxis) {
        val settings = praxis.toMeditationSettings()
        val bgName = resolveBackgroundSoundName(praxis)
        _uiState.update { state ->
            val newMinutes = if (state.timerState == TimerState.Idle) {
                settings.durationMinutes
            } else {
                state.selectedMinutes
            }
            state.copy(
                settings = settings,
                selectedMinutes = newMinutes,
                currentPraxis = praxis,
                resolvedBackgroundSoundName = bgName
            )
        }
    }

    private fun saveSettings() {
        viewModelScope.launch {
            val state = _uiState.value
            praxisRepository.save(
                Praxis.fromMeditationSettings(state.settings, state.currentPraxis.id)
            )
        }
    }

    // MARK: - Lifecycle

    override fun onCleared() {
        super.onCleared()
        timerJob?.cancel()
        // Don't stop service here - let it run if timer is active
    }

    companion object {
        private const val AFFIRMATION_COUNT = 5
    }
}

File: ios/StillMoment/Application/ViewModels/TimerViewModel.swift
//
//  TimerViewModel.swift
//  Still Moment
//
//  Application Layer - Timer ViewModel with Reducer Pattern
//

import Combine
import Foundation
import OSLog

/// ViewModel managing timer state using unidirectional data flow
///
/// This ViewModel uses the Reducer pattern:
/// 1. View dispatches actions
/// 2. Reducer produces effects (pure function)
/// 3. Effects are executed by the effect handler
/// 4. MeditationTimer updates trigger view updates directly
@MainActor
final class TimerViewModel: ObservableObject {
    // MARK: Lifecycle

    init(
        timerService: TimerServiceProtocol = TimerService(),
        audioService: AudioServiceProtocol = AudioService(),
        soundRepository: BackgroundSoundRepositoryProtocol = BackgroundSoundRepository(),
        praxisRepository: PraxisRepository = UserDefaultsPraxisRepository(),
        customAudioRepository: CustomAudioRepositoryProtocol = CustomAudioRepository(),
        soundscapeResolver: SoundscapeResolverProtocol? = nil
    ) {
        self.timerService = timerService
        self.audioService = audioService
        self.soundRepository = soundRepository
        self.praxisRepository = praxisRepository
        self.customAudioRepository = customAudioRepository
        self.soundscapeResolver = soundscapeResolver ?? SoundscapeResolver(
            soundRepository: soundRepository,
            customAudioRepository: customAudioRepository
        )

        let praxis = praxisRepository.load()
        self.currentPraxis = praxis

        // sessionEditor is constructed with a placeholder onSaved because Swift
        // forbids `[weak self]` until every stored property is initialised.
        // The real callback is wired up below, once `self` is fully usable.
        self.sessionEditor = PraxisEditorViewModel(
            praxis: praxis,
            repository: praxisRepository,
            audioService: audioService,
            soundRepository: soundRepository,
            customAudioRepository: customAudioRepository
        ) { _ in }

        self.settings = praxis.toMeditationSettings()
        self.selectedMinutes = self.settings.durationMinutes

        self.setupBindings()

        self.sessionEditor.onSaved = { [weak self] saved in
            self?.updateFromPraxis(saved)
        }
    }

    // MARK: Internal

    // MARK: - Published State

    /// The meditation timer (nil when idle — no dummy object)
    @Published var timer: MeditationTimer?

    /// Selected duration in minutes (1-60), managed directly by the ViewModel
    @Published var selectedMinutes: Int = 10

    /// Meditation settings (interval gongs, background sound, etc.)
    @Published var settings: MeditationSettings = .default

    /// Error message if any operation fails
    @Published var errorMessage: String?

    /// The current Praxis configuration (single config, loaded from repository on init)
    @Published var currentPraxis: Praxis = .default

    /// Editor view model used by the four setting detail views.
    ///
    /// Owned by the TimerViewModel (one instance per session) so the detail
    /// views can be reached without any lazy-init / race-condition gymnastics
    /// in the View layer. Live-save handles persistence; `onSaved` updates
    /// `currentPraxis` here on every change.
    let sessionEditor: PraxisEditorViewModel

    // MARK: - Computed Properties (Forwarded from MeditationTimer)

    /// Current timer state
    var timerState: TimerState {
        self.timer?.state ?? .idle
    }

    /// Remaining time in seconds
    var remainingSeconds: Int {
        self.timer?.remainingSeconds ?? 0
    }

    /// Total duration in seconds
    var totalSeconds: Int {
        self.timer?.totalSeconds ?? 0
    }

    /// Progress value (0.0 - 1.0)
    var progress: Double {
        self.timer?.progress ?? 0.0
    }

    /// Remaining preparation seconds
    var remainingPreparationSeconds: Int {
        self.timer?.remainingPreparationSeconds ?? 0
    }

    /// Whether currently in preparation phase
    var isPreparation: Bool {
        self.timer?.isPreparation ?? false
    }

    /// Visuelle Phase fuer den geteilten `BreathingCircleView`.
    ///
    /// - `.preRoll` waehrend der Vorbereitungs-Countdown laeuft.
    /// - `.playing` ansonsten — der Atemkreis sieht in StartGong, Running und
    ///   EndGong identisch aus (Bogen waechst, Atem laeuft).
    var phase: MeditationPhase {
        self.isPreparation ? .preRoll : .playing
    }

    /// Restzeit fuer das "NOCH … MIN"-Label im Atemkreis-Layout.
    ///
    /// Format `m:ss` (Player-Konvention, ohne Minuten-Padding) — verwendet
    /// zusammen mit `guided_meditations.player.remainingTime.format`.
    var formattedRemainingMinutes: String {
        let minutes = self.remainingSeconds / 60
        let seconds = self.remainingSeconds % 60
        return String(format: "%d:%02d", minutes, seconds)
    }

    /// Returns true if timer can be started
    var canStart: Bool {
        self.timer == nil && self.selectedMinutes > 0
    }

    /// Whether the timer is actively running
    var isRunning: Bool {
        self.timer?.isRunning ?? false
    }

    /// Whether Zen Mode is active: tab bar and distracting UI should be hidden during meditation
    /// Includes the completion screen — tab bar stays hidden until user taps "Back"
    var isZenMode: Bool {
        self.timerState != .idle
    }

    // MARK: - Action Dispatch

    /// Dispatches an action to the reducer and executes resulting effects
    func dispatch(_ action: TimerAction) {
        Logger.viewModel.debug("Dispatching action: \(String(describing: action))")

        let effects = TimerReducer.reduce(
            action: action,
            timerState: self.timer?.state ?? .idle,
            selectedMinutes: self.selectedMinutes,
            settings: self.settings
        )

        self.executeEffects(effects)
    }

    // MARK: - Public Methods

    /// Starts the timer with selected duration
    func startTimer() {
        self.dispatch(.startPressed)
    }

    /// Resets the timer to initial state
    func resetTimer() {
        self.dispatch(.resetPressed)
    }

    /// Applies a Praxis configuration: updates currentPraxis, settings, and selectedMinutes.
    /// Called by `sessionEditor`'s onSaved callback after every live-save.
    func updateFromPraxis(_ praxis: Praxis) {
        self.currentPraxis = praxis
        let settings = praxis.toMeditationSettings()
        self.settings = settings
        self.selectedMinutes = settings.durationMinutes
        Logger.viewModel.info("Updated from praxis")
    }

    // MARK: Private

    private let timerService: TimerServiceProtocol
    let audioService: AudioServiceProtocol
    let soundRepository: BackgroundSoundRepositoryProtocol
    private let praxisRepository: PraxisRepository
    let customAudioRepository: CustomAudioRepositoryProtocol
    let soundscapeResolver: SoundscapeResolverProtocol
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Effect Execution

    private func executeEffects(_ effects: [TimerEffect]) {
        for effect in effects {
            self.executeEffect(effect)
        }
    }

    private func executeEffect(_ effect: TimerEffect) {
        Logger.viewModel.debug("Executing effect: \(String(describing: effect))")

        if self.executeAudioSessionEffect(effect) {
            return
        }
        if self.executeAudioPlaybackEffect(effect) {
            return
        }
        if self.executeTimerEffect(effect) {
            return
        }
    }

    private func executeAudioSessionEffect(_ effect: TimerEffect) -> Bool {
        switch effect {
        case .activateTimerSession:
            self.executeActivateTimerSession()
        case .deactivateTimerSession:
            self.audioService.deactivateTimerSession()
        default:
            return false
        }
        return true
    }

    private func executeAudioPlaybackEffect(_ effect: TimerEffect) -> Bool {
        switch effect {
        case let .startBackgroundAudio(soundId, volume):
            self.executeStartBackgroundAudio(soundId: soundId, volume: volume)
        case .stopBackgroundAudio:
            self.audioService.stopBackgroundAudio()
        case .playStartGong:
            self.executePlayStartGong()
        case let .playIntervalGong(soundId, volume):
            self.executePlayIntervalGong(soundId: soundId, volume: volume)
        case .playCompletionSound:
            self.executePlayCompletionSound()
        default:
            return false
        }
        return true
    }

    private func executeTimerEffect(_ effect: TimerEffect) -> Bool {
        switch effect {
        case let .startTimer(durationMinutes):
            self.executeStartTimer(durationMinutes: durationMinutes)
        case .resetTimer:
            self.timerService.reset()
        case .beginRunningPhase:
            self.timerService.beginRunningPhase()
        case .transitionToCompleted:
            self.timer = self.timer?.withState(.completed)
        case .clearTimer:
            self.timer = nil
        default:
            return false
        }
        return true
    }

    // MARK: - Bindings

    private func setupBindings() {
        self.timerService.timerPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] timer, events in
                self?.handleTimerUpdate(timer, events: events)
            }
            .store(in: &self.cancellables)

        self.audioService.gongCompletionPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] in
                self?.handleGongCompletion()
            }
            .store(in: &self.cancellables)

        self.bindSelectedMinutesToEditor()
    }

    /// Keep the editor's durationMinutes in sync with the wheel picker.
    /// Without this, an editor live-save (e.g. flipping a toggle in a detail
    /// view) would persist the editor's stale durationMinutes and overwrite
    /// whatever the user dialed in on the idle screen.
    private func bindSelectedMinutesToEditor() {
        self.$selectedMinutes
            .removeDuplicates()
            .sink { [weak self] minutes in
                guard let self
                else { return }
                if self.sessionEditor.durationMinutes != minutes {
                    self.sessionEditor.durationMinutes = minutes
                }
            }
            .store(in: &self.cancellables)
    }

    /// Routes gong completion audio callbacks to the appropriate action based on current state
    private func handleGongCompletion() {
        switch self.timerState {
        case .startGong:
            self.dispatch(.startGongFinished)
        case .endGong:
            self.dispatch(.endGongFinished)
        case .running:
            break // Interval gong completed — no state transition needed
        default:
            Logger.viewModel.debug(
                "Gong completion received in unexpected state",
                metadata: ["state": String(describing: self.timerState)]
            )
        }
    }

    private func handleTimerUpdate(_ timer: MeditationTimer, events: [TimerEvent]) {
        // Ignore idle-state timers: TimerService.reset() publishes an idle timer asynchronously
        // via receive(on: DispatchQueue.main), which can arrive after .clearTimer already set
        // self.timer = nil, re-introducing the timer and hiding the Start button.
        guard timer.state != .idle else {
            return
        }

        self.timer = timer

        // Process domain events emitted by tick()
        self.processTimerEvents(events)
    }

    /// Processes domain events from `MeditationTimer.tick()` and dispatches corresponding actions
    private func processTimerEvents(_ events: [TimerEvent]) {
        for event in events {
            switch event {
            case .preparationCompleted:
                Logger.viewModel.info("Meditation starting, dispatching preparationFinished")
                self.dispatch(.preparationFinished)

            case .meditationCompleted:
                Logger.viewModel.info("Timer reached zero, dispatching timerCompleted for endGong phase")
                self.dispatch(.timerCompleted)

            case .intervalGongDue:
                Logger.viewModel.info("Interval gong triggered via domain event")
                self.dispatch(.intervalGongTriggered)
            }
        }
    }
}

// MARK: - Audio Effect Handlers

private extension TimerViewModel {
    func executeActivateTimerSession() {
        do {
            try self.audioService.activateTimerSession()
            Logger.viewModel.info("Timer session activated (audio session + keep-alive)")
        } catch {
            Logger.viewModel.error("Failed to activate timer session", error: error)
            self.errorMessage = "Failed to prepare audio: \(error.localizedDescription)"
        }
    }

    func executeStartBackgroundAudio(soundId: String, volume: Float) {
        do {
            try self.audioService.startBackgroundAudio(soundId: soundId, volume: volume)
        } catch {
            Logger.viewModel.error("Failed to start background audio", error: error)
            self.errorMessage = "Failed to start background audio: \(error.localizedDescription)"
        }
    }

    func executePlayStartGong() {
        do {
            try self.audioService.playStartGong(
                soundId: self.settings.startGongSoundId,
                volume: self.settings.gongVolume
            )
        } catch {
            Logger.viewModel.error("Failed to play start gong", error: error)
            self.errorMessage = "Failed to play start sound: \(error.localizedDescription)"
        }
    }

    func executePlayIntervalGong(soundId: String, volume: Float) {
        do {
            try self.audioService.playIntervalGong(soundId: soundId, volume: volume)
        } catch {
            Logger.viewModel.error("Failed to play interval gong", error: error)
            self.errorMessage = "Failed to play interval sound: \(error.localizedDescription)"
        }
    }

    func executePlayCompletionSound() {
        do {
            try self.audioService.playCompletionSound(
                soundId: self.settings.startGongSoundId,
                volume: self.settings.gongVolume
            )
        } catch {
            Logger.viewModel.error("Failed to play completion sound", error: error)
            self.errorMessage = "Failed to play sound: \(error.localizedDescription)"
        }
    }
}

// MARK: - Timer Effect Handlers

private extension TimerViewModel {
    func executeStartTimer(durationMinutes: Int) {
        self.settings.durationMinutes = durationMinutes
        // Persist the selected duration so the next launch restores it
        let updatedPraxis = self.currentPraxis.withDurationMinutes(durationMinutes)
        self.currentPraxis = updatedPraxis
        self.praxisRepository.save(updatedPraxis)
        // Use preparation time from settings, or 0 if disabled
        let preparationTime = self.settings.preparationTimeEnabled
            ? self.settings.preparationTimeSeconds
            : 0

        // Build interval settings from current meditation settings (nil when disabled)
        let intervalSettings: IntervalSettings? = self.settings.intervalGongsEnabled
            ? IntervalSettings(
                intervalMinutes: self.settings.intervalMinutes,
                mode: self.settings.intervalMode
            )
            : nil

        self.timerService.start(
            durationMinutes: durationMinutes,
            preparationTimeSeconds: preparationTime,
            intervalSettings: intervalSettings
        )
    }
}

// Audio Preview and SwiftUI Preview support: see TimerViewModel+Preview.swift
// Configuration description labels: see TimerViewModel+ConfigurationDescription.swift
// (access level widened for cross-file extension use)
